Preview: Lambda-based constraints

gunnarmorling opened the pull request:
Hey @gsmet, @emmanuelbernard That's a preview how Lambda-based constraints could look like as proposed in [BVAL-515](https://hibernate.atlassian.net/browse/BVAL-515). Check out the test for the actual usage. Some thoughts from my side:

* I kinda like how it optimizes for the simplest case (no initialize() needed as the constraint has no parameters; no access to constraint validator context needed)
* Should we provide an alternative for where these things are needed (or leave that to implementing an actual class)?
* It doesn't pose a compatability issues as originally was David's concern, as we just add new types but don't alter `ConstraintValidator`

On the implementation, I've added it to our DSL to get quickly going, in BV it'd likely be on `Configuration` or some other defined place in the bootstrap API. And don't get hung up on the usage of ByteBuddy, it was just the quickest thing for me to use it, I think it can be implemented without bytecode generation (needs some changes in HV though which expects actual class types for validators)

I think it'd make sense to have this as some sort of prototype API first in HV 6 (Alpha/Beta) and then add it to the spec. Thoughts?

gunnarmorling commented on engine/src/test/java/org/hibernate/validator/test/cfg/ConstraintDefinitionTest.java:
```diff
@@ -269,6 +271,32 @@ public void testMultipleValidatorsForSameTypeWithNoCallToIncludeExistingValidato
+	@Test
+	public void testConstraintBasedOnLambdaExpression() {
+		mapping.constraintDefinition( Directory.class )
+				.validateType( File.class ).with( b -> b.exists() );
```
...or a Lambda.

gsmet commented:
> I kinda like how it optimizes for the simplest case (no initialize() needed as the constraint has no parameters; no access to constraint validator context needed)

+1

> Should we provide an alternative for where these things are needed (or leave that to implementing an actual class)?

I don't think so. Let's keep it simple for the simple case and implement a class for further needs.

> I think it can be implemented without bytecode generation

Yes, it would be nice to avoid it if we can.

> I think it'd make sense to have this as some sort of prototype API first in HV 6 (Alpha/Beta) and then add it to the spec.

+1, I think people needs to play with it in real life cases so that we can have useful feedback.

hferentschik commented:
So what's the idea really, provide an alternative/new constraint mapping API? I guess from that perspective it might be ok, but it in the end BV has always been about declarative constraint mapping. Also, I find it odd that this is just working for simple cases. I would expect something which works for all cases. If not, I'd rather stick to what we have. Lambadas are the new generics and will get overused as generics did.

gunnarmorling commented:
Hey Hardy,

> So what's the idea really, provide an alternative/new constraint mapping API?

No, not a mapping API really. Rather a simplified way to implement a constraint's logic. It means to address user requests about making constraint implementation more light-weight.

> BV has always been about declarative constraint mapping. 

That wouldn't change, you'd still put your constraints in a declarative fashion.

> I would expect something which works for all cases

Yes, that's a question I have, too. I can see how we could have two overloaded methods or so. But it'd violate the "provide one way to do a thing" paradigm. In that light I kinda like to provide just a more light-weight approach for the most common case.

> Lambadas are the new generics and will get overused as generics did.

Hah, don't be that pessimistic. I believe in the good in people :)

I think we should add it in one way or another as experimental in HV 6 Alpha1 and look for feedback.

hferentschik commented:
> That wouldn't change, you'd still put your constraints in a declarative fashion.

Ok

> Yes, that's a question I have, too. I can see how we could have two overloaded methods or so. But it'd violate the "provide one way to do a thing" paradigm. In that light I kinda like to provide just a more light-weight approach for the most common case.

As said, I am not a big fan, especially if it just addresses one case.

> I think we should add it in one way or another as experimental in HV 6 Alpha1 and look for feedback.

I guess I could also see it as some experimental feature to explore it, but I don't think it is ready for BV.

gunnarmorling commented:
Hey @gsmet I've pushed an update, it's not using ByteBuddy anymore now. Let me know what you think. From my side it's good (apart from doc update).

gunnarmorling commented:
@gsmet Can we merge it?

gsmet commented on engine/src/main/java/org/hibernate/validator/cfg/context/ConstraintDefinitionContext.java:
```diff
@@ -39,4 +39,27 @@
+
+		/**
+		 * Applies the given Lambda expression or referenced method to values to be validated. It is guaranteed that
+		 * never {@code null} is passed to these expressions or methods.
```
"{@code null} is never passed" would be better?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/cfg/context/ConstraintDefinitionContextImpl.java:
```diff
@@ -32,7 +32,7 @@
 	private boolean includeExistingValidators = true;
 
-	private final Set<Class<? extends ConstraintValidator<A, ?>>> validatorTypes = newHashSet();
+	private final Set<ConstraintValidatorDescriptor<A>> validatorTypes = new HashSet<>();
```
I don't think `validatorTypes` is accurate as the descriptor also contains the lambda in the lambda case. Why not `validatorDescriptors`?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintdefinition/ConstraintDefinitionContribution.java:
```diff
@@ -18,14 +19,14 @@
 public class ConstraintDefinitionContribution<A extends Annotation> {
 	private final Class<A> constraintType;
-	private final List<Class<? extends ConstraintValidator<A, ?>>> constraintValidators = new ArrayList<Class<? extends ConstraintValidator<A, ?>>>();
+	private final List<ConstraintValidatorDescriptor<A>> validatorTypes;
```
Same here.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ClassBasedValidatorDescriptor.java:
```diff
@@ -0,0 +1,74 @@
+
+	@Override
+	public Type getValidatedType() {
+		return TypeHelper.extractType( validatorClass );
```
Wouldn't it be better to do this in the constructor and have a final variable with it?

gsmet replied:
Btw, just thinking out loud, but any chance we could use ClassMate instead of exposing a new `TypeHelper` method?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ClassBasedValidatorDescriptor.java:
```diff
@@ -0,0 +1,74 @@
+
+	@Override
+	public EnumSet<ValidationTarget> getValidationTargets() {
+		SupportedValidationTarget supportedTargetAnnotation = validatorClass.getAnnotation(
```
Same here?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ConstraintValidatorDescriptor.java:
```diff
@@ -0,0 +1,39 @@
+import org.hibernate.validator.cfg.context.ConstraintDefinitionContext.ValidationCallable;
+
+/**
+ * Represents a specific validator type.
```
Not sure the Javadoc is accurate (see my remark about `validatorTypes` earlier).

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ConstraintValidatorDescriptor.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public interface ConstraintValidatorDescriptor<A extends Annotation> extends Serializable {
+
+	Class<? extends ConstraintValidator<A, ?>> getValidatorClass();
```
Maybe a little more formatting and javadoc love is in order?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ConstraintValidatorManager.java:
```diff
@@ -115,13 +115,13 @@ public ConstraintValidatorManager(ConstraintValidatorFactory constraintValidator
 		}
 
-		Class<? extends ConstraintValidator<?, ?>> validatorClass = findMatchingValidatorClass(
+		ConstraintValidatorDescriptor<A> validatorTypeDescriptor = findMatchingValidatorClass(
```
`validatorDescriptor` would be nice to be consistent with the naming a few lines below.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/metadata/core/ConstraintHelper.java:
```diff
@@ -238,19 +241,28 @@ public ConstraintHelper() {
 
-	private static <A extends Annotation> void putConstraint(Map<Class<? extends Annotation>, List<? extends Class<?>>> validators, Class<A> constraintType, Class<? extends ConstraintValidator<A, ?>> validatorType) {
-		validators.put( constraintType, Collections.singletonList( validatorType ) );
+	private static <A extends Annotation> void putConstraint(Map<Class<? extends Annotation>, List<ConstraintValidatorDescriptor<?>>> validators, Class<A> constraintType, Class<? extends ConstraintValidator<A, ?>> validatorType) {
```
Good to have something more type safe now.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/metadata/core/ConstraintHelper.java:
```diff
@@ -320,39 +332,35 @@ private boolean isBuiltinConstraint(Class<? extends Annotation> annotationType)
 	 *
 	 * @param annotationType The constraint annotation type
-	 * @param definitionClasses The validators to register
+	 * @param validatorTypes The validator types to register
```
Same remark about validator types.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/metadata/descriptor/ConstraintDescriptorImpl.java:
```diff
@@ -256,23 +260,14 @@ public ConstraintTarget getValidationAppliesTo() {
 	 * @return The validators applying to type of this constraint.
 	 */
-	public List<Class<? extends ConstraintValidator<T, ?>>> getMatchingConstraintValidatorClasses() {
+	public List<ConstraintValidatorDescriptor<T>> getMatchingConstraintValidatorClasses() {
```
Shouldn't we change the name to be clear it return descriptors?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/util/TypeHelper.java:
```diff
@@ -295,25 +297,25 @@ private static Type getResolvedSuperclass(Type type) {
 	 *         key is the type the validator accepts and value the validator class itself.
 	 */
-	public static <A extends Annotation> Map<Type, Class<? extends ConstraintValidator<A, ?>>> getValidatorsTypes(
+	public static <A extends Annotation> Map<Type, ConstraintValidatorDescriptor<A>> getValidatorTypes(
```
Same remark about validator types not being accurate.

gsmet replied:
Looks like you missed this one.

gunnarmorling commented:
@gsmet Thanks for reviewing. I've pushed some more commits for

* Addressing your remarks about the naming (in a way "type" could be applied to the descriptors, too, but I agree that "class" was too far off. Should be fixed everywhere now
* Adding a new annotation `@Incubating` (see [HV-1141](https://hibernate.atlassian.net/projects/HV/issues/HV-1141)) and putting it to the new API members
* Adding docs

gunnarmorling commented:
Taking a step back, there are two loose ends remaining:

* currently `validateType` only takes a `Class` object, but I think we should have one for `Type`, too. Because a user may wish to define an expression for `List<String>` instead of raw `List` (e.g. an equivalent to `SizeValidatorForCollection`)
* currently there is no way to declare a cross-parameter constraint using a Lambda expression (as the supported validation target(s) is controlled via an annotation on `ConstraintValidator` implementations)

I think we can do these two separately as there always is the current way of doing things.

gsmet commented on documentation/src/main/asciidoc/ch11.asciidoc:
```diff
@@ -6,6 +6,13 @@ in addition to the functionality defined by the Bean Validation specification. T
 
+New APIs or SPIs are tagged with the `org.hibernate.validator.Incubating` annotation as long as they
+are under development. This means that such elements (e.g. packages, types, methods, constants etc.)
+may be incompatible altered - or removed - in subsequent releases. Usage of incubating API/SPI members
```
incompatibly?

gsmet commented on engine/src/main/java/org/hibernate/validator/Incubating.java:
```diff
@@ -0,0 +1,23 @@
+
+/**
+ * Marks the annotated element as incubating. The contract of incubating elements (e.g. packages, types, methods,
+ * constants etc.) is under active development and may be incompatible altered - or removed - in subsequent releases.
```
incompatibly?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ClassBasedValidatorDescriptor.java:
```diff
@@ -0,0 +1,81 @@
+			return EnumSet.copyOf( Arrays.asList( supportedTargetAnnotation.value() ) );
+		}
+	}
+	@Override
```
Missing new line just above.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ClassBasedValidatorDescriptor.java:
```diff
@@ -0,0 +1,81 @@
+
+	public ClassBasedValidatorDescriptor(Class<? extends ConstraintValidator<A, ?>> validatorClass) {
+		this.validatorClass = validatorClass;
+		this.validatedType = TypeHelper.extractType( validatorClass );
```
Did you see my comment about avoiding exposing a new TypeHelper method and use ClassMate instead? Maybe it does not make sense here and has a risk of not being consistent with the rest but I thought I might as well ask.

gsmet commented:
> I think we can do these two separately as there always is the current way of doing things.

Could you create JIRA issues for them?

gsmet commented:
@gunnarmorling I added a few more comments on your changes. I can fix them on merging if you prefer!

gunnarmorling commented:
Sweet, thanks! I'll rework my last commits to address your final remarks and then do the push.

gunnarmorling commented:
> Could you create JIRA issues for them?

Done: [HV-1142](https://hibernate.atlassian.net/browse/HV-1142).

gunnarmorling commented:
Merged it, yippie :)
